src/OrchardCore.Modules/OrchardCore.Roles/Services/RoleUpdater.cs

```diff
@@ -47,68 +56,109 @@ public class RoleUpdater : IFeatureEventHandler
+        /// </summary>
+        /// <returns></returns>
+        private async Task AssignPermissionsToRolesAsync()
         {
-            // when another module is being enabled, locate matching permission providers
-            var providersForEnabledModule = _permissionProviders
-                .Where(x => _typeFeatureProvider.GetFeatureForDependency(x.GetType()).Id == feature.Id);
+            if (_updateInProgress)
+            {
+                return;
+            }
+
+            _updateInProgress = true;
+
+            var roleNames = await _roleService.GetRoleNamesAsync();
+
+            if (!roleNames.Any())
+            {
+                // Site roles are initially added using the "RolesStep" handler, while no role names are availble. This means
+                // that RoleUpdater handler was called before the "RolesStep".
+                // Its likley to be coming from a tenant setup request. Nothing to do.
+                return;
+            }
+
+            var rolesDocument = await _rolesDocumentManager.GetOrCreateMutableAsync();
+            var updateRolesDocument = false;
+
+            // Get all the available permissions grouped by role name as defined by IPermissionProvider.
+            var groups = _permissionProviders.SelectMany(x => x.GetDefaultStereotypes())
+                .GroupBy(stereotype => stereotype.Name)
+                .Select(x => new
+                {
+                    RoleName = x.Key,
+                    PermissionNames = x.SelectMany(y => y.Permissions ?? Enumerable.Empty<Permission>()).Select(x => x.Name)
+                });
 
-            if (_logger.IsEnabled(LogLevel.Debug))
+            foreach (var group in groups)
             {
-                if (providersForEnabledModule.Any())
+                if (!roleNames.Any(roleName => roleName.Equals(group.RoleName, StringComparison.OrdinalIgnoreCase))
+                    || await _roleManager.FindByNameAsync(group.RoleName) is not Role role)
                 {
-                    _logger.LogDebug("Configuring default roles for feature '{FeatureName}'", feature.Id);
+                    // A role is mapped in IPermissionProvider, yet it isn't available for the tenant, ignore it.
+                    continue;
                 }
-                else
+
+                var currentPermissionNames = role.RoleClaims.Where(x => x.ClaimType == Permission.ClaimType).Select(x => x.ClaimValue);
+
+                var distinctPermissionNames = currentPermissionNames
+                    .Union(group.PermissionNames)
+                    .Distinct()
+                    .ToList();
+
+                if (!rolesDocument.PermissionGroups.ContainsKey(group.RoleName))
                 {
-                    _logger.LogDebug("No default roles for feature '{FeatureName}'", feature.Id);
+                    rolesDocument.PermissionGroups.TryAdd(group.RoleName, new List<string>());
```

Comment:
If you want to replace th List then use `[group.RoleName] = ...` and if you don't want to, keep using `TryAdd` and remove the `ContainsKey` usage